File: src/java/magmac/app/error/ImmutableCompileError.java
package magmac.app.error;

import magmac.api.Tuple2;
import magmac.api.collect.list.List;
import magmac.api.collect.list.Lists;
import magmac.api.iter.collect.Joiner;
import magmac.api.iter.collect.Max;
import magmac.app.compile.error.context.Context;
import magmac.app.compile.error.error.CompileError;

public record ImmutableCompileError(String message, Context context,
                                    List<CompileError> errors) implements CompileError {
    public ImmutableCompileError(String message, Context context) {
        this(message, context, Lists.empty());
    }

    private static String formatEntry(int depth, String display) {
        var repeated = "| ".repeat(depth + 1);
        return "\n" + repeated + display;
    }

    private static String joinSorted(int depth, List<CompileError> copy, List<Integer> indices) {
        return copy.iterWithIndices()
                .map((tuple) -> {
                    var indices1 = ImmutableCompileError.getIntegerList(copy, indices, tuple);
                    return tuple.right().format(depth + 1, indices1);
                })
                .map((String display) -> ImmutableCompileError.formatEntry(depth, display))
                .collect(new Joiner())
                .orElse("");
    }

    private static List<Integer> getIntegerList(List<CompileError> copy, List<Integer> indices, Tuple2<Integer, CompileError> tuple) {
        if (1 >= copy.size()) {
            return indices;
        }
        else {
            return indices.addLast(tuple.left());
        }
    }

    @Override
    public String display() {
        return this.format(0, Lists.empty());
    }

    @Override
    public String format(int depth, List<Integer> indices) {
        var copy = this.errors.sort((CompileError first, CompileError second) -> first.computeMaxDepth() - second.computeMaxDepth());
        var joined = ImmutableCompileError.joinSorted(depth, copy, indices);

        var joinedIndices = indices.iter()
                .map(String::valueOf)
                .collect(new Joiner("."))
                .orElse("");

        return joinedIndices + ") " + this.message + ": " + this.context.display() + joined;
    }

    @Override
    public int computeMaxDepth() {
        var compileErrorIter = this.errors.iter();
        return 1 + compileErrorIter.map(CompileError::computeMaxDepth)
                .collect(new Max())
                .orElse(0);
    }
}

File: src/web/magmac/app/error/ImmutableCompileError.ts
import { Tuple2 } from "../../../magmac/api/Tuple2";
import { List } from "../../../magmac/api/collect/list/List";
import { Lists } from "../../../magmac/api/collect/list/Lists";
import { Joiner } from "../../../magmac/api/iter/collect/Joiner";
import { Max } from "../../../magmac/api/iter/collect/Max";
import { Context } from "../../../magmac/app/compile/error/context/Context";
import { CompileError } from "../../../magmac/app/compile/error/error/CompileError";
export class ImmutableCompileError {
	 ImmutableCompileError( message : String,  context : Context) : public {this( message, context, Lists.empty( ));;}
	private static formatEntry( depth : int,  display : String) : String { let repeated : var="| ".repeat( depth+1);return "\n"+repeated+display;;}
	private static joinSorted( depth : int,  copy : List<CompileError>,  indices : List<Integer>) : String {return copy.iterWithIndices( ).map( 0).map( 0).collect( new Joiner( )).orElse( "");;}
	private static getIntegerList( copy : List<CompileError>,  indices : List<Integer>,  tuple : Tuple2<Integer, CompileError>) : List<Integer> {if(true){ return indices;;}if(true){ return indices.addLast( tuple.left( ));;};}
	public display() : String {return this.format( 0, Lists.empty( ));;}
	public format( depth : int,  indices : List<Integer>) : String { let copy : var=this.errors.sort( 0); let joined : var=ImmutableCompileError.joinSorted( depth, copy, indices); let joinedIndices : var=indices.iter( ).map( String.valueOf).collect( new Joiner( ".")).orElse( "");return joinedIndices+") "+this.message+": "+this.context.display( )+joined;;}
	public computeMaxDepth() : int { let compileErrorIter : var=this.errors.iter( );return 1+compileErrorIter.map( CompileError.computeMaxDepth).collect( new Max( )).orElse( 0);;}
}
